[4.0] Improve user token plugin display in the profile page

nikosdion opened the pull request:
Pull Request for Issue #29338 and #29337

cc @PhilETaylor @Quy 

### Summary of Changes

Changed the way the Joomla Token user profile field renders its information in the User Profile view page.

### Testing Instructions

Set up a token and view – NOT edit! – your user profile in the public frontend of the site.

### Expected result

Under Joomla API Token you get reasonable information e.g. what's your token and whether it's enabled.

### Actual result

You get all sorts of misleading information. The token field displays the seed, not the token you need to use. The Active field displays 1 / 0 which great if you are a 🤖. And of course there's the Reset field which makes no sense in the context.

This PR fixes this by displaying only the _real_ token you need to use and whether it's enabled. Example:
![image](https://user-images.githubusercontent.com/256041/83966351-62a00f00-a8c2-11ea-8353-66dfee113d31.png)

### Documentation Changes Required

None.

Commit:
Improve user token plugin display in the profile page

Closes gh-29338
Closes gh-29337

SharkyKZ commented:
The form is also altered when editing.

SharkyKZ commented on plugins/user/token/token.php:
```diff
@@ -150,15 +151,42 @@ public function onContentPrepareData(string $context, &$data): bool
+		 * generic and we run the risk of creating naming clashes. Instead, we manipulate the data
+		 * directly.
+		 */
+		if ($context = 'com_users.profile')
```
```suggestion
		if ($context === 'com_users.profile')
```

Commit:
Update plugins/user/token/token.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

nikosdion commented:
@SharkyKZ Thanks. I think I now understand what happened when I pressed backspace and my browser stayed on the same page. My keyboard was focused on the wrong window...

SharkyKZ commented:
No problem. But this does not fix the issue mentioned above. The form uses `com_users.profile` name both when editing and displaying so this won't work.

Commit:
Well, Joomla is using the same context for different things

nikosdion commented:
That's annoying so let's do it the same way I've been doing it in my own software: checking the layout in the request. Architecturally wrong but necessary as it seems 😞

richard67 commented:
I have tested this item :white_check_mark: successfully on 20b973bf78e3683ec78dd98534f041bb5e213c52<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29469">issues.joomla.org/tracker/joomla-cms/29469</a>.</sub>

Quy commented:
Reset option is missing in edit mode.

richard67 commented:
Oh, seems I've missed that.

richard67 commented:
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29469">issues.joomla.org/tracker/joomla-cms/29469</a>.</sub>

wilsonge commented on plugins/user/token/token.php:
```diff
@@ -151,14 +152,41 @@ public function onContentPrepareData(string $context, &$data): bool
+		 * generic and we run the risk of creating naming clashes. Instead, we manipulate the data
+		 * directly.
+		 */
+		if (($context === 'com_users.profile') && (Factory::getApplication()->input->get('layout') !== 'edit'))
```
```suggestion
		if (($context === 'com_users.profile') && ($this->app->input->get('layout') !== 'edit'))
```

wilsonge commented on plugins/user/token/token.php:
```diff
@@ -251,6 +279,12 @@ public function onContentPrepareForm(Form $form, $data): bool
 		}
 
+		// Remove the Reset field when displaying the user profile form
+		if ($form->getName() == 'com_users.profile')
```
I think we probably want to check the layout here too? Because I guess when editing the profile this field does make sense?

nikosdion replied:
Yup. That's what @Quy was saying. This bug predates this PR but since I got into the trouble of fixing it let me do it properly :)

wilsonge commented on plugins/user/token/token.php:
```diff
@@ -528,4 +560,74 @@ private function isInAllowedUserGroup($userId)
+			return $algo;
+		}
+
+		if (preg_match('/\s*algo=\s*"\s*([a-z0-9]+)\s*"/i', $contents, $matches) !== 1)
```
Is reading this out with SimpleXMLElement not a better match here? As we're in `onContentPrepareForm` you already have access to the parsed XML anyway in JForm so there's not even a performance impact of reparsing the file I don't think - this is presumably intended to be a property on a named field

nikosdion replied:
SimpleXMLElement is much slower than this RegEx (about 4 times slower). JForm is even slower. Also, since onContentPrepareData is called before onContentPrepareForm we don't have access to the Form object of the user profile. Therefore we'd have to have SimpleXMLElement parse the same form XML file twice: once in onContentPrepareData and once on onContentPrepareForm (through JForm). Yeah, I really thought about it before using a RegEx, that's why I documented why I did that 14 lines above, in the method's docblock ;)

wilsonge replied:
🤦 and i actually made sure to check the event it was being called from. and still got it wrong :D I must have been on my A-Game :)

wilsonge commented on plugins/user/token/token.php:
```diff
@@ -528,4 +560,74 @@ private function isInAllowedUserGroup($userId)
+
+		try
+		{
+			$siteSecret = Factory::getApplication()->get('secret');
```
```suggestion
			$siteSecret = $this->app->get('secret');
```

wilsonge commented on plugins/user/token/token.php:
```diff
@@ -528,4 +560,74 @@ private function isInAllowedUserGroup($userId)
+		$tokenHash = hash_hmac($algorithm, $rawToken, $siteSecret);
+		$message   = base64_encode("$algorithm:$userId:$tokenHash");
+
+		if ($userId !== Factory::getUser()->id)
```
```suggestion
		if ($userId !== $this->app->getIdentity()->id)
```

SharkyKZ replied:
And cast it as integer.

nikosdion replied:
As an aside, there is no need to cast as an integer. I am using `!==` which tests for type equality _and_ value equality. Moreover, `$userId` is guaranteed to be an integer since there's a scalar typehint. As for User's `id` property it's always an integer. If it's a NULL or any other non-integer value it's not a valid identity and the test needs to fail which it does because of the type-safe comparison operator.

SharkyKZ replied:
>As for User's id property it's always an integer.

It can be a string when using PDO driver.

nikosdion replied:
It must NEVER be a string. Ever. If it's a string there's something wrong in the User object. That field must always be either an integer or null. If using the PDO driver results in a string you need to make a core change to typecast the id into an int or null when the data is bound to the object. This is why we're using a user _object_ instead of a user _array_.

Commit:
Update plugins/user/token/token.php

Co-authored-by: George Wilson <georgejameswilson@googlemail.com>

Commit:
Update plugins/user/token/token.php

Co-authored-by: George Wilson <georgejameswilson@googlemail.com>

Commit:
Update plugins/user/token/token.php

Co-authored-by: George Wilson <georgejameswilson@googlemail.com>

Commit:
Display the token reset in the frontend profile page

Commit:
Token would be reset if saving the user without the profile fields

Backported fix from FOF 3 token plugin, the code this core Joomla
feature was forked from when I originally contributed it.

Commit:
Reformat docblock, remove reference to FOF

nikosdion commented:
OK, this is ready for testing now.

I also fixed a bug I discovered about the token after I originally contributed this feature. If you saved the user object _without_ passing user profile fields (or at least without passing the token profile fields) the token would be automatically reset. I had already fixed it in the FOF Token plugin  which is where this core feature was forked from but forgot to file a PR to Joomla. Since I'm here fixing this feature I backported this bugfix from FOF as well.

Quy commented:
I have tested this item :white_check_mark: successfully on 8d76b1ce6b9eb4e8f79813b1c7e736e82f1fe208<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29469">issues.joomla.org/tracker/joomla-cms/29469</a>.</sub>

richard67 commented:
I have tested this item :white_check_mark: successfully on 8d76b1ce6b9eb4e8f79813b1c7e736e82f1fe208<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29469">issues.joomla.org/tracker/joomla-cms/29469</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29469">issues.joomla.org/tracker/joomla-cms/29469</a>.</sub>

wilsonge commented:
Thanks Nic!
